### PR TITLE
Documentation for GraphFeaturePreprocessor

### PR DESCRIPTION
**Pull request by kubilay-atasu**

### CODE REVIEW

File: docs/source/graph_preprocessor.rst [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+The GraphFeaturePreprocessor is a scikit-learn compatible preprocessing library that enables feature extraction from graph-structured data.  
+The library includes utilities for creating, updating, and transforming graph-structured data. 
+The basic input and output data representation used by the library is an edge list with edge features.
+As an example, in a financial transaction graph, the graph nodes represent accounts and the graph edges represent transactions.
+The date of the transactions as well as the amount transferred between the accounts may be provided as edge features.
+However, the GraphFeaturePreprocessor can generate a much richer set of edge features by analyzing the graph topology.   
+For instance, the GraphFeaturePreprocessor can extract the number of cycles each transaction is a part of in the graph.
+Such additional features can significantly increase the predictive power of machine learning models in practical settings. 
+
+.. image:: graph-feature-extraction.png
```
Comment (andreea-anghel) [inline docs/source/graph_preprocessor.rst:13]: Very nice! Just a few comments: I would bold the title and do another screenshot without the red lines under set_params and XGBoost.

File: docs/source/graph_preprocessor.rst [hunk 2]
```diff
@@ -0,0 +1,41 @@
+Graph-Based Feature Extraction
+==============================
+
+The GraphFeaturePreprocessor is a scikit-learn compatible preprocessing library that enables feature extraction from graph-structured data.
```
Comment (andreea-anghel) [inline docs/source/graph_preprocessor.rst:4]: I would avoid saying that the graph preprocessor is a library, because we only have one library now, namely Snap ML. 

I would say something like (please feel free to adapt it as you see fit): "The GraphFeaturePreprocessor is a scikit-learn compatible pre-processor for graph-based data. It provides a set of functions that can enrich raw feature vectors with meaningful graph-related patterns. "

File: docs/source/graph_preprocessor.rst [hunk 3]
```diff
@@ -0,0 +1,41 @@
+Graph-Based Feature Extraction
+==============================
+
+The GraphFeaturePreprocessor is a scikit-learn compatible preprocessing library that enables feature extraction from graph-structured data.  
+The library includes utilities for creating, updating, and transforming graph-structured data.
```
Comment (andreea-anghel) [inline docs/source/graph_preprocessor.rst:5]: same comment about the library. I would say the GraphFeaturePreprocessor provides utilities for...

File: docs/source/graph_preprocessor.rst [hunk 4]
```diff
@@ -0,0 +1,41 @@
+Graph-Based Feature Extraction
+==============================
+
+The GraphFeaturePreprocessor is a scikit-learn compatible preprocessing library that enables feature extraction from graph-structured data.  
+The library includes utilities for creating, updating, and transforming graph-structured data. 
+The basic input and output data representation used by the library is an edge list with edge features.
```
Comment (andreea-anghel) [inline docs/source/graph_preprocessor.rst:6]: I would add a screenshot with an example of such data (a table with a few transactions and columns and column names. It's ok to say edge list with edge features, but not sure all people understand this terminology/format.

Comment (andreea-anghel) [reply docs/source/graph_preprocessor.rst:6]: ok, ignore this comment please, just saw you added it a table below, great.

File: docs/source/graph_preprocessor.rst [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+| 2000              | 1                 | 4                 | 20                |                  |                   |
++-------------------+-------------------+-------------------+-------------------+------------------+-------------------+
+| 4000              | 3                 | 2                 | 30                |                  |                   |
++-------------------+-------------------+-------------------+-------------------+------------------+-------------------+
+
+The edge list must be sorted in the ascending order of the timestamps for best performance. 
+An in-memory directed multi-graph representation will be created internally when an edge list is provided as input to the `fit` function.
+The `transform` function, on the other hand, receives an edge list as input, inserts these edges into the in-memory graph,
+and extracts new features from the graph, which are appended to the input edge list as additional columns to form the output of this function.   
+
```
Comment (andreea-anghel) [inline docs/source/graph_preprocessor.rst:40]: would it make sense to add also a code snippet that shows the sequence of the steps described above?  something like:

```
from snapml import GraphFeaturePreprocessor

# Instatiate a graph feature preprocessor object
gp = GraphFeaturePreprocessor()

# define the configuration of the graph preprocessor
params = {
    "num_threads": 12,  # number of software threads to be used
    "time_window": 10 * 3600,  # time window used if no pattern was specified
    "scatter-gather": True,
    "scatter-gather_tw": 120 * 3600,
    "scatter-gather_bins": [ 2, 3 ],
}

# set the parameters of the greaph feature preprocessor
gp.set_params(params)

# load old transactions
many_old_transactions = np.loadtxt(old_transactions_file_path)

# Create the graph out of the transactions loaded above
gp.fit(many_old_transactions)

# load an unknown transaction, here X_test comprises one transaction
new_transaction = np.loadtxt(new_transaction_file_path)
enriched_new_transaction = gp.transform(new_transaction)
```

Comment (kubilay-atasu) [reply docs/source/graph_preprocessor.rst:40]: I did not see code snippets in other rest files. Don't we use examples for that purpose?

Comment (andreea-anghel) [reply docs/source/graph_preprocessor.rst:40]: True, the examples repository is the main place where we store the code examples. However, if it will take long to have the example ready (e.g., due to datasets not being publicly available), then we could (temporarily) provide at least this code snippet in the documentation. What do you think?

**Comment by kubilay-atasu** [comment]
@andreea-anghel: there is already some data and examples here:
https://github.ibm.com/ai-foundation/snapml/tree/jov-graph-preproc/test/graph-preprocessor/examples

**Comment by andreea-anghel** [comment]
@kubilay-atasu yes, those could be used, but is it ok to release the aml_e_small dataset used by these examples? if yes, then no need for a code snippet in the documentation.

**Comment by kubilay-atasu** [comment]
@andreea-anghel I see that both the dataset and the example scripts using it are part of the PR you have created:
https://github.ibm.com/ai-foundation/snapml/pull/625

Did you plan to remove them?

**Comment by kubilay-atasu** [comment]
There is also a very small dataset I had created for testing the features extracted. See aml_custom_* here:
https://github.ibm.com/ai-foundation/snapml/tree/jov-graph-preproc/test/graph-preprocessor/input

We are using this dataset already in the unittests:
https://github.ibm.com/ai-foundation/snapml/blob/jov-graph-preproc/test/graph-preprocessor/UnitTests.py
class GraphFeaturesCustomTest(unittest.TestCase):

**Comment by kubilay-atasu** [comment]
The custom dataset is not created using Erik Altman's simulator. We can use it as we wish.

**Comment by andreea-anghel** [comment]
> @andreea-anghel I see that both the dataset and the example scripts using it are part of the PR you have created: https://github.ibm.com/ai-foundation/snapml/pull/625
> 
> Did you plan to remove them?

No. Those examples are not public anywhere, just stored in the SnapML repository for development/testing. The public examples will be available here https://github.com/IBM/snapml-examples/tree/main/examples/training. Here will go the notebook that I will create based on the examples provided in the SnapML repository.

**Comment by andreea-anghel** [comment]
> The custom dataset is not created using Erik Altman's simulator. We can use it as we wish.

@kubilay-atasu Ok, great! Then let's use this custom dataset. I will create the notebook based on this one. And we don't add any more code snippets to the documentation. Would you agree?

**Comment by kubilay-atasu** [comment]
Agreed, but note that the custom dataset is only for testing the feature extraction, and not for inference. There are no labels, and it is too small to train anything on it anyway.

**Comment by andreea-anghel** [comment]
> Agreed, but note that the custom dataset is only for testing the feature extraction, and not for inference. There are no labels, and it is too small to train anything on it anyway.

Absolutely, just to show how to use API of the new Snap ML feature. 

Forperformance, we will need to wait and see what big dataset to use publicly.

**Comment by andreea-anghel** [comment]
please do not merge until we have the new feature out, thanks!
